app/user.go

```diff
@@ -52,29 +101,101 @@ func (a *App) Pay(to string, amount uint64) error {
 		},
 	}
 
-	// The transaction must be signed.
-	if txn, err := tbody.Sign(a.CurrentUser.Wallet, crand.Reader); err == nil {
-
-		// The transaction must be broadcasted to the peers.
-		if err := wallet.SetPending(txn); err != nil {
-			return err
-		}
+	// Any change left over gets sent back to the sender
+	if totalInput > amount {
+		tbody.Outputs = append(tbody.Outputs, blockchain.TxOutput{
+			Amount:    totalInput - amount,
+			Recipient: wallet.Public().Repr(),
+		})
+	}
 
+	// The transaction must be signed.
+	if txn, err := tbody.Sign(*a.CurrentUser.Wallet, crand.Reader); err == nil {
 		// The transaction must be added to the pool.
 		code := pool.Push(txn, a.Chain)
 		if code != consensus.ValidTransaction {
-			return errors.New("transaction validation failed")
+			return fmt.Errorf("Transaction validation failed with code %d", code)
+		}
+
+		// The transaction must be added to the wallet's pending transcations
+		if err := wallet.SetPending(txn); err != nil {
+			return err
 		}
 
 		// The transaction must be broadcasted to the network.
 		a.PeerStore.Broadcast(msg.Push{
 			ResourceType: msg.ResourceTransaction,
 			Resource:     txn,
 		})
-
 	} else {
 		return err
 	}
-
 	return nil
 }
+
+// collectInputsForTxn returns a list of input transactions for a new transaction
+// from the given sender of the given amount, and the total value of all the
+// inputs returned. Returns an error if there are not enough transactions to the
+// given sender in the blockchain to make a new transaction of the given amount.
+func (a *App) collectInputsForTxn(sender string, amount uint64) ([]blockchain.TxHashPointer,
+	uint64, error) {
+
+	a.Chain.RLock()
+	defer a.Chain.RUnlock()
+
+	total := uint64(0)
+	inputs := make([]blockchain.TxHashPointer, 0)
+	spentTxns := make(map[blockchain.Hash]*blockchain.Transaction, 0)
+
+	// Iterate through the blockchain in reverse, searching for transactions to
+	// sender.
+	for i := len(a.Chain.Blocks) - 1; i >= 0; i-- {
+		block := a.Chain.Blocks[i]
+		txns := block.GetTransactionsTo(sender)
+
+		// Add all the transactions to sender to our list of potential input
+		// transactions until the total is greater that or equal to the amount
+		// for the transaction we want to send.
+		for i, txn := range txns {
+			if i != 0 {
+				// This transaction is not a CloudBase transaction. Add it's
+				// inputs to a map of spent transactions so we don't spend them.
+				for _, hashPtr := range txn.Inputs {
```

Comment:
Maybe only add spent TXNs to the map if they _could_ be used by us. The proportion of transactions to any given user will be small